source/containerimage/ocilayout.go

```diff
@@ -0,0 +1,195 @@
+package containerimage
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/reference"
+	"github.com/containerd/containerd/remotes"
+	"github.com/moby/buildkit/session"
+	sessioncontent "github.com/moby/buildkit/session/content"
+	"github.com/moby/buildkit/source"
+	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// GetOCILayoutResolver gets a resolver to an OCI layout for a specified scope from the pool
+func GetOCILayoutResolver(layoutPath string, sm *session.Manager, sessionID string, g session.Group) *OCILayoutResolver {
+	r := &OCILayoutResolver{
+		path:      layoutPath,
+		sm:        sm,
+		sessionID: sessionID,
+		g:         g,
+	}
+	return r
+}
+
+type OCILayoutResolver struct {
+	remotes.Resolver
+	path      string
+	sm        *session.Manager
+	sessionID string
+	g         session.Group
+	is        images.Store
+	mode      source.ResolveMode
+}
+
+// WithImageStore returns new resolver that can also resolve from local images store
+func (r *OCILayoutResolver) WithImageStore(is images.Store, mode source.ResolveMode) *OCILayoutResolver {
+	r2 := *r
+	r2.is = is
+	r2.mode = mode
+	return &r2
+}
+
+// Fetcher returns a new fetcher for the provided reference.
+func (r *OCILayoutResolver) Fetcher(ctx context.Context, ref string) (remotes.Fetcher, error) {
+	return r, nil
+}
+
+// Fetch get an io.ReadCloser for the specific content
+func (r *OCILayoutResolver) Fetch(ctx context.Context, desc ocispecs.Descriptor) (io.ReadCloser, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	sessionID := r.sessionID
+
+	caller, err := r.sm.Get(timeoutCtx, sessionID, false)
+	if err != nil {
+		return r.fetchWithAnySession(ctx, desc)
+	}
+
+	return r.fetchWithSession(ctx, desc, caller)
+}
+
+func (r *OCILayoutResolver) fetchWithAnySession(ctx context.Context, desc ocispecs.Descriptor) (io.ReadCloser, error) {
+	var rc io.ReadCloser
+	err := r.sm.Any(ctx, r.g, func(ctx context.Context, _ string, caller session.Caller) error {
+		readCloser, err := r.fetchWithSession(ctx, desc, caller)
+		if err != nil {
+			return err
+		}
+		rc = readCloser
+		return nil
+	})
+	return rc, err
+}
+
+func (r *OCILayoutResolver) fetchWithSession(ctx context.Context, desc ocispecs.Descriptor, caller session.Caller) (io.ReadCloser, error) {
+	store := sessioncontent.NewCallerStore(caller, fmt.Sprintf("oci-layout:%s", r.path))
+	readerAt, err := store.ReaderAt(ctx, desc)
+	if err != nil {
+		return nil, err
+	}
+	// just wrap the ReaderAt with a Reader
```

Comment:
`content.ReaderAt` implements closer.

There is also a `Seeker` check in this `io.ReadCloser` that should be implemented if you want partial uploads to continue cleanly.

Reply:
> content.ReaderAt implements closer.

So it does. But of course, it does not implement `Read()`, so it still needs a wrapper. Fixed.

Reply:
> There is also a Seeker check in this io.ReadCloser that should be implemented if you want partial uploads to continue cleanly.

I didn't get that. Uploads?